src/backends/llvm.ts

```diff
@@ -1,33 +1,232 @@
+		
+		let [lType, _1] = emitter.ir.type_table[tree.lhs.id!];
+		let [rType, _2] = emitter.ir.type_table[tree.rhs.id!];
+
+		if (lType === INT && rType === INT) {
+			switch (tree.op) {
+				case "+": {
+					return emitter.builder.add(v1, v2, "addtmp");
+				}
+				case "*": {
+					return emitter.builder.mul(v1, v2, "multmp");
+				}
+				default: {
+					throw "Unknown bin op";
+				}
+			}
+		} else if (lType === FLOAT && rType === FLOAT) {
+			switch (tree.op) {
+				case "+": {
+					return emitter.builder.addf(v1, v2, "addtmp");
+				}
+				case "*": {
+					return emitter.builder.mulf(v1, v2, "multmp");
+				}
+				default: {
+					throw "Unknown bin op";
+				}
+			}
+		} else if (lType === FLOAT && rType === INT) {
+			// TODO
+			throw "Not implemented yet"
+		}
+		else if (lType === INT && rType === FLOAT) {
+			// TODO
+			throw "Not implemented yet"
+		} else {
+			throw "Incompatible Operands";
+		}
+	},
+
+	visit_quote(tree: ast.QuoteNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_escape(tree: ast.EscapeNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_run(tree: ast.RunNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_fun(tree: ast.FunNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_call(tree: ast.CallNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_extern(tree: ast.ExternNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_persist(tree: ast.PersistNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_if(tree: ast.IfNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_while(tree: ast.WhileNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	},
+
+	visit_macrocall(tree: ast.MacroCallNode, emitter: LLVMEmitter): llvm.Value {
+		throw "not implemented";
+	}
 };
 
-export function compile(tree: ast.SyntaxNode, emitter: LLVMEmitter): string {
-  return ast_visit(compile_rules, tree, emitter);
+export function compile(tree: ast.SyntaxNode, emitter: LLVMEmitter): llvm.Value {
+	return ast_visit(compile_rules, tree, emitter);
 }
 
 // Compile the IR to a complete JavaScript program.
 export function codegen(ir: CompilerIR): string {
-  let emitter: LLVMEmitter = {
-    ir: ir,
-    compile: compile,
-    emit_proc: emit_proc,
-    emit_prog: emit_prog,
-    emit_prog_variant: emit_prog_variant,
-    variant: null,
-    builder: null,
-  };
-
-  // Emit and invoke the main (anonymous) function.
-  return emit_main_wrapper(emit_main(emitter));
+	let emitter: LLVMEmitter = {
+		ir: ir,
+		compile: compile,
+		emit_proc: emit_proc,
+		emit_prog: emit_prog,
+		emit_prog_variant: emit_prog_variant,
+		variant: null,
+		builder: null,
+	};
+
+	// Emit and invoke the main (anonymous) function.
+	return emit_main_wrapper(emit_main(emitter));
 }
```

Comment:
We may want to coordinate our indentation styles. :smiley: I have my editor configured to use two spaces, which I think is pretty standard in TS code—you seem to have it configured to use "real" tabs. I try to avoid having strong opinions on these things, but for the sake of consistency with what's currently in the Braid repository, would you mind at least temporarily switching to spaces?